compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/BridgeLowering.kt

```diff
@@ -1,287 +1,382 @@
-        signaturesToSkip: MutableSet<Method>,
-        specialOverrideInfo: SpecialMethodWithDefaultInfo?,
-        isSpecial: Boolean
-    ) {
-        val signature = method.getJvmSignature()
-        if (signature in signaturesToSkip) return
-
-        val bridge = createBridgeHeader(irClass, target, method, isSpecial = isSpecial, isSynthetic = !isSpecial)
-        bridge.createBridgeBody(target, specialOverrideInfo, isSpecial)
-        irClass.declarations.add(bridge)
-
-        // For lambda classes, we move override from the `invoke` function to its bridge. This will allow us to avoid boxing
-        // the return type of `invoke` in codegen, in case lambda's return type is primitive.
-        if (method.name == OperatorNameConventions.INVOKE && irClass.origin == JvmLoweredDeclarationOrigin.LAMBDA_IMPL) {
-            target.overriddenSymbols.remove(method.symbol)
-            bridge.overriddenSymbols.add(method.symbol)
+    // Returns the special bridge overridden by the current methods if it exists.
+    private val IrSimpleFunction.specialBridgeOrNull: SpecialBridge?
+        get() = specialBridgeCache.getOrPutNullable(symbol) {
+            val specialMethodInfo = specialBridgeMethods.getSpecialMethodInfo(this)
+            return when {
+                specialMethodInfo != null ->
+                    // Note that there are type-safe special bridges with generic return types in Java (namely Map.getOrDefault,
+                    // Map.get, and MutableMap.remove), but the JVM backend does not produce overrides with specialized return
+                    // types for them. So for compatibility, neither do we.
+                    SpecialBridge(this, jvmMethod, methodInfo = specialMethodInfo)
+
+                specialBridgeMethods.isBuiltInWithDifferentJvmName(this) ->
+                    if (returnType.isTypeParameter())
+                        SpecialBridge(this, jvmMethod, specializedReturnType = returnType)
+                    else
+                        SpecialBridge(this, jvmMethod)
+
+                else -> {
+                    val overriddenSpecialBridge = overriddenSymbols.asSequence().mapNotNull { it.owner.specialBridgeOrNull }.firstOrNull()
+                    if (overriddenSpecialBridge?.specializedReturnType != null) {
+                        val specializedSignature = Method(
+                            overriddenSpecialBridge.signature.name,
+                            context.methodSignatureMapper.mapReturnType(this),
+                            overriddenSpecialBridge.signature.argumentTypes
+                        )
+                        overriddenSpecialBridge.copy(signature = specializedSignature, specializedReturnType = returnType)
+                    } else {
+                        overriddenSpecialBridge
+                    }
+                }
+            }
         }
-
-        signaturesToSkip.add(signature)
+    private val specialBridgeMethods = SpecialBridgeMethods(context)
```

Comment:
In this case it's probably fine as is, but could do with a bit of refactoring.

The `SpecialBridgeMethods` don't keep any state (just a reference to the backend context) and should be refactored into a collection of top-level functions.

The `specialBridgeCache` is just there to avoid quadratic behavior in `overriddenFinalSpecialBridges`. It would probably be cleaner to just refactor this, too.